{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4676bbc0a60fd540",
   "metadata": {},
   "source": [
    "# Download data from cellxgene census"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4a9863c",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install -U cellxgene-census"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36dde4169145c358",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:22.463639Z",
     "start_time": "2026-01-17T21:36:18.982912Z"
    }
   },
   "outputs": [],
   "source": [
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma(census_version=\"latest\")\n",
    "census[\"census_info\"][\"summary\"].read().concat().to_pandas()\n",
    "\n",
    "# Get all dataset citations from census\n",
    "datasets = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1edfe0fa28cd9eff",
   "metadata": {},
   "source": [
    "## Fetch AnnData from Census"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7870b1ff57b782bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch an AnnData object\n",
    "dataset_id = \"df1edb87-e512-43ae-b5f4-cb179cfc2bb4\" # https://datasets.cellxgene.cziscience.com/31380664-ba9c-49d1-9961-b2bf4f7131a2.h5ad\n",
    "cellxgene_census.download_source_h5ad(dataset_id, to_path=\"/content/raw-data/cxg_sjogrens.h5ad\", progress_bar=True)\n",
    "census.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbf34a72555641f",
   "metadata": {},
   "source": [
    "## Fetch citation for dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d80ff1553c1ca83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a citation string for the slice\n",
    "datasets[datasets[\"dataset_id\"] == dataset_id].iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb3b824738bee6d",
   "metadata": {},
   "source": [
    "# Prepare data for Geneformer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd8b7c35cf9263a",
   "metadata": {},
   "source": [
    "#### The input data for Geneformer is single-cell RNA-seq (scRNAseq) data composed of a raw count matrix stored in .loom or .h5ad format. If data is in another format (e.g., .csv, .mtx), it should be converted to .loom or .h5ad using Scanpy or Anndata tools.\n",
    "\n",
    "#### No normalization, transformation, or feature selection should be performed. Removal of cells per standard quality controls (e.g., removal of empty droplets, damaged cells, doublets, etc.) is recommended."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c5f0fceea7e3888",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:41.155470Z",
     "start_time": "2026-01-17T21:36:34.492529Z"
    }
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "adata = sc.read_h5ad(\"/content/raw-data/cxg_sjogrens.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9769583420272a5c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:41.201462Z",
     "start_time": "2026-01-17T21:36:41.157368Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 94227 × 31969\n",
       "    obs: 'n_genes', 'n_counts', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_rp', 'pct_counts_rp', 'total_counts_hb', 'pct_counts_hb', 'Sample', 'batch', 'leiden', 'leiden_before', 'Immune_All_High.pkl', 'Immune_All_AddPIP.pkl', 'Immune_All_Low.pkl', 'Immune_All_PIP.pkl', 'Cells_Intestinal_Tract.pkl', 'Cells_Lung_Airway.pkl', 'Developing_Mouse_Brain.pkl', 'Nuclei_Lung_Airway.pkl', 'Pan_Fetal_Human.pkl', 'Focus Score', 'SSA', 'ANA', 'Annotation', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'development_stage_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'medical_background', 'suspension_type', 'sex_ontology_term_id', 'donor_id', 'is_primary_data', 'tissue_type', 'cell_type', 'assay', 'disease', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'mt', 'rp', 'hb', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'feature_type'\n",
       "    uns: 'Cells_Lung_Airway.pkl_colors', 'Developing_Mouse_Brain.pkl_colors', 'Immune_All_AddPIP.pkl_colors', 'Immune_All_High.pkl_colors', 'Immune_All_Low.pkl_colors', 'Immune_All_PIP.pkl_colors', 'Nuclei_Lung_Airway.pkl_colors', 'Sample_colors', 'batch_condition', 'citation', 'default_embedding', 'dendrogram_leiden', 'hvg', 'leiden', 'leiden_colors', 'neighbors', 'organism', 'organism_ontology_term_id', 'pca', 'rank_genes_groups', 'schema_reference', 'schema_version', 'tissue_ontology_term_id_colors', 'title', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c2b5648023db7c7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:41.257508Z",
     "start_time": "2026-01-17T21:36:41.206069Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>total_counts_mt</th>\n",
       "      <th>pct_counts_mt</th>\n",
       "      <th>total_counts_rp</th>\n",
       "      <th>pct_counts_rp</th>\n",
       "      <th>total_counts_hb</th>\n",
       "      <th>pct_counts_hb</th>\n",
       "      <th>...</th>\n",
       "      <th>is_primary_data</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>assay</th>\n",
       "      <th>disease</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>observation_joinid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCAATTCTCTTGGTCC-1-1</th>\n",
       "      <td>7669</td>\n",
       "      <td>238340.0</td>\n",
       "      <td>7669</td>\n",
       "      <td>238340.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.030629</td>\n",
       "      <td>27903.0</td>\n",
       "      <td>11.707225</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>acinar cell of salivary gland</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>African American</td>\n",
       "      <td>35-year-old stage</td>\n",
       "      <td>p*esn;7zns</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAGATTGTCACGAACT-1-1</th>\n",
       "      <td>7949</td>\n",
       "      <td>236529.0</td>\n",
       "      <td>7949</td>\n",
       "      <td>236529.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.015220</td>\n",
       "      <td>38970.0</td>\n",
       "      <td>16.475780</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>acinar cell of salivary gland</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>African American</td>\n",
       "      <td>35-year-old stage</td>\n",
       "      <td>B~?5gyw$k8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGATCTTTCTTGGTCC-1-1</th>\n",
       "      <td>6614</td>\n",
       "      <td>221995.0</td>\n",
       "      <td>6614</td>\n",
       "      <td>221995.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>39420.0</td>\n",
       "      <td>17.757156</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>acinar cell of salivary gland</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>African American</td>\n",
       "      <td>35-year-old stage</td>\n",
       "      <td>0kYbSzkgca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCGTTCAAGACATATG-1-1</th>\n",
       "      <td>7965</td>\n",
       "      <td>221002.0</td>\n",
       "      <td>7965</td>\n",
       "      <td>221002.0</td>\n",
       "      <td>1875.0</td>\n",
       "      <td>0.848409</td>\n",
       "      <td>27994.0</td>\n",
       "      <td>12.666854</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000452</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>myoepithelial cell</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>African American</td>\n",
       "      <td>35-year-old stage</td>\n",
       "      <td>6CN0&amp;gx15+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGTTATCATCGAGCC-1-1</th>\n",
       "      <td>7341</td>\n",
       "      <td>219529.0</td>\n",
       "      <td>7341</td>\n",
       "      <td>219529.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>38020.0</td>\n",
       "      <td>17.318897</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>acinar cell of salivary gland</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>normal</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>African American</td>\n",
       "      <td>35-year-old stage</td>\n",
       "      <td>&lt;p*%{_&gt;GJr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TACTTGTGTCTACAAC-1-38</th>\n",
       "      <td>154</td>\n",
       "      <td>218.0</td>\n",
       "      <td>154</td>\n",
       "      <td>218.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.917431</td>\n",
       "      <td>28.0</td>\n",
       "      <td>12.844036</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.458716</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>acinar cell of salivary gland</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>Sjogren syndrome</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>European American</td>\n",
       "      <td>50-year-old stage</td>\n",
       "      <td>X95IfDnuo&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATGGTTGTCGGAGATG-1-38</th>\n",
       "      <td>74</td>\n",
       "      <td>129.0</td>\n",
       "      <td>74</td>\n",
       "      <td>129.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.550388</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.325581</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>CD8-positive, alpha-beta cytotoxic T cell</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>Sjogren syndrome</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>European American</td>\n",
       "      <td>50-year-old stage</td>\n",
       "      <td>a{*oSt0NK#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCGGTTCGCGGACT-1-38</th>\n",
       "      <td>164</td>\n",
       "      <td>188.0</td>\n",
       "      <td>164</td>\n",
       "      <td>188.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.063830</td>\n",
       "      <td>11.0</td>\n",
       "      <td>5.851064</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>Sjogren syndrome</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>European American</td>\n",
       "      <td>50-year-old stage</td>\n",
       "      <td>918pjw$xmw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACCTTTGTTCATCTT-1-38</th>\n",
       "      <td>100</td>\n",
       "      <td>124.0</td>\n",
       "      <td>100</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.419355</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>CD8-positive, alpha-beta cytotoxic T cell</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>Sjogren syndrome</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>European American</td>\n",
       "      <td>50-year-old stage</td>\n",
       "      <td>S3S|}CpDJN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATCGTATCGACGTCG-1-38</th>\n",
       "      <td>101</td>\n",
       "      <td>127.0</td>\n",
       "      <td>101</td>\n",
       "      <td>127.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.362205</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>tissue</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "      <td>10x 3' v3</td>\n",
       "      <td>Sjogren syndrome</td>\n",
       "      <td>female</td>\n",
       "      <td>labial gland</td>\n",
       "      <td>European American</td>\n",
       "      <td>50-year-old stage</td>\n",
       "      <td>CtXhVT9qy6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94227 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       n_genes  n_counts  n_genes_by_counts  total_counts  \\\n",
       "TCAATTCTCTTGGTCC-1-1      7669  238340.0               7669      238340.0   \n",
       "CAGATTGTCACGAACT-1-1      7949  236529.0               7949      236529.0   \n",
       "TGATCTTTCTTGGTCC-1-1      6614  221995.0               6614      221995.0   \n",
       "CCGTTCAAGACATATG-1-1      7965  221002.0               7965      221002.0   \n",
       "GGGTTATCATCGAGCC-1-1      7341  219529.0               7341      219529.0   \n",
       "...                        ...       ...                ...           ...   \n",
       "TACTTGTGTCTACAAC-1-38      154     218.0                154         218.0   \n",
       "ATGGTTGTCGGAGATG-1-38       74     129.0                 74         129.0   \n",
       "TTCCGGTTCGCGGACT-1-38      164     188.0                164         188.0   \n",
       "AACCTTTGTTCATCTT-1-38      100     124.0                100         124.0   \n",
       "GATCGTATCGACGTCG-1-38      101     127.0                101         127.0   \n",
       "\n",
       "                       total_counts_mt  pct_counts_mt  total_counts_rp  \\\n",
       "TCAATTCTCTTGGTCC-1-1              73.0       0.030629          27903.0   \n",
       "CAGATTGTCACGAACT-1-1              36.0       0.015220          38970.0   \n",
       "TGATCTTTCTTGGTCC-1-1               0.0       0.000000          39420.0   \n",
       "CCGTTCAAGACATATG-1-1            1875.0       0.848409          27994.0   \n",
       "GGGTTATCATCGAGCC-1-1               0.0       0.000000          38020.0   \n",
       "...                                ...            ...              ...   \n",
       "TACTTGTGTCTACAAC-1-38              2.0       0.917431             28.0   \n",
       "ATGGTTGTCGGAGATG-1-38              2.0       1.550388              3.0   \n",
       "TTCCGGTTCGCGGACT-1-38              2.0       1.063830             11.0   \n",
       "AACCTTTGTTCATCTT-1-38              0.0       0.000000              3.0   \n",
       "GATCGTATCGACGTCG-1-38              0.0       0.000000              3.0   \n",
       "\n",
       "                       pct_counts_rp  total_counts_hb  pct_counts_hb  ...  \\\n",
       "TCAATTCTCTTGGTCC-1-1       11.707225              0.0       0.000000  ...   \n",
       "CAGATTGTCACGAACT-1-1       16.475780              0.0       0.000000  ...   \n",
       "TGATCTTTCTTGGTCC-1-1       17.757156              0.0       0.000000  ...   \n",
       "CCGTTCAAGACATATG-1-1       12.666854              1.0       0.000452  ...   \n",
       "GGGTTATCATCGAGCC-1-1       17.318897              0.0       0.000000  ...   \n",
       "...                              ...              ...            ...  ...   \n",
       "TACTTGTGTCTACAAC-1-38      12.844036              1.0       0.458716  ...   \n",
       "ATGGTTGTCGGAGATG-1-38       2.325581              0.0       0.000000  ...   \n",
       "TTCCGGTTCGCGGACT-1-38       5.851064              0.0       0.000000  ...   \n",
       "AACCTTTGTTCATCTT-1-38       2.419355              0.0       0.000000  ...   \n",
       "GATCGTATCGACGTCG-1-38       2.362205              0.0       0.000000  ...   \n",
       "\n",
       "                      is_primary_data tissue_type  \\\n",
       "TCAATTCTCTTGGTCC-1-1             True      tissue   \n",
       "CAGATTGTCACGAACT-1-1             True      tissue   \n",
       "TGATCTTTCTTGGTCC-1-1             True      tissue   \n",
       "CCGTTCAAGACATATG-1-1             True      tissue   \n",
       "GGGTTATCATCGAGCC-1-1             True      tissue   \n",
       "...                               ...         ...   \n",
       "TACTTGTGTCTACAAC-1-38            True      tissue   \n",
       "ATGGTTGTCGGAGATG-1-38            True      tissue   \n",
       "TTCCGGTTCGCGGACT-1-38            True      tissue   \n",
       "AACCTTTGTTCATCTT-1-38            True      tissue   \n",
       "GATCGTATCGACGTCG-1-38            True      tissue   \n",
       "\n",
       "                                                       cell_type      assay  \\\n",
       "TCAATTCTCTTGGTCC-1-1               acinar cell of salivary gland  10x 3' v3   \n",
       "CAGATTGTCACGAACT-1-1               acinar cell of salivary gland  10x 3' v3   \n",
       "TGATCTTTCTTGGTCC-1-1               acinar cell of salivary gland  10x 3' v3   \n",
       "CCGTTCAAGACATATG-1-1                          myoepithelial cell  10x 3' v3   \n",
       "GGGTTATCATCGAGCC-1-1               acinar cell of salivary gland  10x 3' v3   \n",
       "...                                                          ...        ...   \n",
       "TACTTGTGTCTACAAC-1-38              acinar cell of salivary gland  10x 3' v3   \n",
       "ATGGTTGTCGGAGATG-1-38  CD8-positive, alpha-beta cytotoxic T cell  10x 3' v3   \n",
       "TTCCGGTTCGCGGACT-1-38                                 fibroblast  10x 3' v3   \n",
       "AACCTTTGTTCATCTT-1-38  CD8-positive, alpha-beta cytotoxic T cell  10x 3' v3   \n",
       "GATCGTATCGACGTCG-1-38            CD4-positive, alpha-beta T cell  10x 3' v3   \n",
       "\n",
       "                                disease     sex        tissue  \\\n",
       "TCAATTCTCTTGGTCC-1-1             normal  female  labial gland   \n",
       "CAGATTGTCACGAACT-1-1             normal  female  labial gland   \n",
       "TGATCTTTCTTGGTCC-1-1             normal  female  labial gland   \n",
       "CCGTTCAAGACATATG-1-1             normal  female  labial gland   \n",
       "GGGTTATCATCGAGCC-1-1             normal  female  labial gland   \n",
       "...                                 ...     ...           ...   \n",
       "TACTTGTGTCTACAAC-1-38  Sjogren syndrome  female  labial gland   \n",
       "ATGGTTGTCGGAGATG-1-38  Sjogren syndrome  female  labial gland   \n",
       "TTCCGGTTCGCGGACT-1-38  Sjogren syndrome  female  labial gland   \n",
       "AACCTTTGTTCATCTT-1-38  Sjogren syndrome  female  labial gland   \n",
       "GATCGTATCGACGTCG-1-38  Sjogren syndrome  female  labial gland   \n",
       "\n",
       "                      self_reported_ethnicity  development_stage  \\\n",
       "TCAATTCTCTTGGTCC-1-1         African American  35-year-old stage   \n",
       "CAGATTGTCACGAACT-1-1         African American  35-year-old stage   \n",
       "TGATCTTTCTTGGTCC-1-1         African American  35-year-old stage   \n",
       "CCGTTCAAGACATATG-1-1         African American  35-year-old stage   \n",
       "GGGTTATCATCGAGCC-1-1         African American  35-year-old stage   \n",
       "...                                       ...                ...   \n",
       "TACTTGTGTCTACAAC-1-38       European American  50-year-old stage   \n",
       "ATGGTTGTCGGAGATG-1-38       European American  50-year-old stage   \n",
       "TTCCGGTTCGCGGACT-1-38       European American  50-year-old stage   \n",
       "AACCTTTGTTCATCTT-1-38       European American  50-year-old stage   \n",
       "GATCGTATCGACGTCG-1-38       European American  50-year-old stage   \n",
       "\n",
       "                      observation_joinid  \n",
       "TCAATTCTCTTGGTCC-1-1          p*esn;7zns  \n",
       "CAGATTGTCACGAACT-1-1          B~?5gyw$k8  \n",
       "TGATCTTTCTTGGTCC-1-1          0kYbSzkgca  \n",
       "CCGTTCAAGACATATG-1-1          6CN0&gx15+  \n",
       "GGGTTATCATCGAGCC-1-1          <p*%{_>GJr  \n",
       "...                                  ...  \n",
       "TACTTGTGTCTACAAC-1-38         X95IfDnuo>  \n",
       "ATGGTTGTCGGAGATG-1-38         a{*oSt0NK#  \n",
       "TTCCGGTTCGCGGACT-1-38         918pjw$xmw  \n",
       "AACCTTTGTTCATCTT-1-38         S3S|}CpDJN  \n",
       "GATCGTATCGACGTCG-1-38         CtXhVT9qy6  \n",
       "\n",
       "[94227 rows x 47 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af6d1c9e985986a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:41.437103Z",
     "start_time": "2026-01-17T21:36:41.302856Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mt</th>\n",
       "      <th>rp</th>\n",
       "      <th>hb</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>feature_is_filtered</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_reference</th>\n",
       "      <th>feature_biotype</th>\n",
       "      <th>feature_length</th>\n",
       "      <th>feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5.936808e-06</td>\n",
       "      <td>-0.580874</td>\n",
       "      <td>-3.119813</td>\n",
       "      <td>False</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>517</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1015</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>2618</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>4.593860e-05</td>\n",
       "      <td>0.493031</td>\n",
       "      <td>-1.894246</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1319</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239906</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000239906</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>323</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>3.515507e-02</td>\n",
       "      <td>6.892033</td>\n",
       "      <td>5.408454</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>294</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>2.517536e-02</td>\n",
       "      <td>3.115924</td>\n",
       "      <td>1.099064</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>351</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>3.006966e-02</td>\n",
       "      <td>2.679734</td>\n",
       "      <td>0.601274</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>2922</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>831</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>510</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31969 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    mt     rp     hb  highly_variable         means  \\\n",
       "ENSG00000243485  False  False  False            False  5.936808e-06   \n",
       "ENSG00000237613  False  False  False            False  1.000000e-12   \n",
       "ENSG00000186092  False  False  False            False  1.000000e-12   \n",
       "ENSG00000239945  False  False  False            False  4.593860e-05   \n",
       "ENSG00000239906  False  False  False            False  1.000000e-12   \n",
       "...                ...    ...    ...              ...           ...   \n",
       "ENSG00000277856  False  False  False             True  3.515507e-02   \n",
       "ENSG00000275063  False  False  False             True  2.517536e-02   \n",
       "ENSG00000271254  False  False  False             True  3.006966e-02   \n",
       "ENSG00000277475  False  False  False            False  1.000000e-12   \n",
       "ENSG00000268674  False  False  False            False  1.000000e-12   \n",
       "\n",
       "                 dispersions  dispersions_norm  feature_is_filtered  \\\n",
       "ENSG00000243485    -0.580874         -3.119813                False   \n",
       "ENSG00000237613          NaN          0.000000                False   \n",
       "ENSG00000186092          NaN          0.000000                False   \n",
       "ENSG00000239945     0.493031         -1.894246                False   \n",
       "ENSG00000239906          NaN          0.000000                False   \n",
       "...                      ...               ...                  ...   \n",
       "ENSG00000277856     6.892033          5.408454                False   \n",
       "ENSG00000275063     3.115924          1.099064                False   \n",
       "ENSG00000271254     2.679734          0.601274                False   \n",
       "ENSG00000277475          NaN          0.000000                False   \n",
       "ENSG00000268674          NaN          0.000000                False   \n",
       "\n",
       "                    feature_name feature_reference feature_biotype  \\\n",
       "ENSG00000243485      MIR1302-2HG    NCBITaxon:9606            gene   \n",
       "ENSG00000237613          FAM138A    NCBITaxon:9606            gene   \n",
       "ENSG00000186092            OR4F5    NCBITaxon:9606            gene   \n",
       "ENSG00000239945  ENSG00000239945    NCBITaxon:9606            gene   \n",
       "ENSG00000239906  ENSG00000239906    NCBITaxon:9606            gene   \n",
       "...                          ...               ...             ...   \n",
       "ENSG00000277856  ENSG00000277856    NCBITaxon:9606            gene   \n",
       "ENSG00000275063  ENSG00000275063    NCBITaxon:9606            gene   \n",
       "ENSG00000271254  ENSG00000271254    NCBITaxon:9606            gene   \n",
       "ENSG00000277475  ENSG00000277475    NCBITaxon:9606            gene   \n",
       "ENSG00000268674  ENSG00000268674    NCBITaxon:9606            gene   \n",
       "\n",
       "                feature_length    feature_type  \n",
       "ENSG00000243485            517          lncRNA  \n",
       "ENSG00000237613           1015          lncRNA  \n",
       "ENSG00000186092           2618  protein_coding  \n",
       "ENSG00000239945           1319          lncRNA  \n",
       "ENSG00000239906            323          lncRNA  \n",
       "...                        ...             ...  \n",
       "ENSG00000277856            294  protein_coding  \n",
       "ENSG00000275063            351  protein_coding  \n",
       "ENSG00000271254           2922  protein_coding  \n",
       "ENSG00000277475            831  protein_coding  \n",
       "ENSG00000268674            510  protein_coding  \n",
       "\n",
       "[31969 rows x 13 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "558fe20334f5a0d2",
   "metadata": {},
   "source": [
    "#### The genes must be labeled as Ensembl IDs. If your dataset includes only gene symbols instead, you can use external tools like Ensembl Biomart or MyGeneInfo to convert them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fe570e1d1eac5dc9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:41.803273Z",
     "start_time": "2026-01-17T21:36:41.747593Z"
    }
   },
   "outputs": [],
   "source": [
    "# Duplicate index to a new .var column -> \"ensembl_id\"\n",
    "adata.var[\"ensembl_id\"] = adata.var.index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33529ab407bfc724",
   "metadata": {},
   "source": [
    "#### The data may contain additional feature labels that one would like to retain in the tokenized dataset, for example to use as labels for embeddings or fine-tuning, or to confirm balancing of attributes across data splits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "80468a5fc4fdaed2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:42.342348Z",
     "start_time": "2026-01-17T21:36:42.313079Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Sjogren syndrome', 'normal'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sjogren syndrome\n",
    "# normal (control)\n",
    "set(adata.obs[\"disease\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e540b4fdf6dbadff",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:42.461456Z",
     "start_time": "2026-01-17T21:36:42.429021Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'CD4-positive, alpha-beta T cell',\n",
       " 'CD8-positive, alpha-beta cytotoxic T cell',\n",
       " 'CD8-positive, alpha-beta regulatory T cell',\n",
       " 'IgA plasma cell',\n",
       " 'IgG plasma cell',\n",
       " 'acinar cell of salivary gland',\n",
       " 'alternatively activated macrophage',\n",
       " 'dendritic cell',\n",
       " 'duct epithelial cell',\n",
       " 'effector CD8-positive, alpha-beta T cell',\n",
       " 'endothelial cell',\n",
       " 'fibroblast',\n",
       " 'inflammatory macrophage',\n",
       " 'ionocyte',\n",
       " 'mature NK T cell',\n",
       " 'myoepithelial cell',\n",
       " 'pro-T cell',\n",
       " 'smooth muscle cell'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cell type annotations from original authors\n",
    "set(adata.obs[\"cell_type\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8a17fed54cda6de8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:42.550360Z",
     "start_time": "2026-01-17T21:36:42.493013Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'CD4-positive, alpha-beta T cell',\n",
       " 'CD8-positive, alpha-beta cytotoxic T cell',\n",
       " 'CD8-positive, alpha-beta regulatory T cell',\n",
       " 'Macrophage',\n",
       " 'Plasma cell',\n",
       " 'acinar cell of salivary gland',\n",
       " 'dendritic cell',\n",
       " 'duct epithelial cell',\n",
       " 'effector CD8-positive, alpha-beta T cell',\n",
       " 'endothelial cell',\n",
       " 'fibroblast',\n",
       " 'ionocyte',\n",
       " 'mature NK T cell',\n",
       " 'myoepithelial cell',\n",
       " 'pro-T cell',\n",
       " 'smooth muscle cell'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# collapse macrophage and plasma cell subtypes into broader categories\n",
    "rename_dict = {\n",
    "    \"alternatively activated macrophage\": \"Macrophage\",\n",
    "    \"inflammatory macrophage\": \"Macrophage\",\n",
    "    \"IgA plasma cell\": \"Plasma cell\",\n",
    "    \"IgG plasma cell\": \"Plasma cell\"\n",
    "}\n",
    "\n",
    "# create unified cell-type annotation\n",
    "adata.obs[\"cell_type_union\"] = [\n",
    "    rename_dict.get(ct, ct) for ct in adata.obs[\"cell_type\"]\n",
    "]\n",
    "\n",
    "# sanity check\n",
    "set(adata.obs[\"cell_type_union\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "27a164614219d8db",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:42.599395Z",
     "start_time": "2026-01-17T21:36:42.552103Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'35',\n",
       " '38',\n",
       " '39',\n",
       " '40',\n",
       " '43',\n",
       " '45',\n",
       " '47',\n",
       " '48',\n",
       " '50',\n",
       " '51',\n",
       " '53',\n",
       " '54',\n",
       " '55',\n",
       " '56',\n",
       " '58',\n",
       " '59',\n",
       " '61',\n",
       " '65',\n",
       " '75'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reformat developmental stage / age attribute\n",
    "set(adata.obs[\"development_stage\"])\n",
    "\n",
    "# keep only first two characters (e.g., \"Adult\" -> \"Ad\")\n",
    "adata.obs[\"development_stage\"] = [\n",
    "    ds[:2] for ds in adata.obs[\"development_stage\"]\n",
    "]\n",
    "\n",
    "# sanity check\n",
    "set(adata.obs[\"development_stage\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "588227c08ce7ba1b",
   "metadata": {},
   "source": [
    "#### Optionally, one may provide a binary \"filter_pass\" attribute that indicates whether to include (1) or exclude (0) the cell from tokenization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dea033513f5fe0fc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:42.660850Z",
     "start_time": "2026-01-17T21:36:42.605096Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "filter_pass\n",
       "0    57375\n",
       "1    36852\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select cell types relevant to complement production, regulation, and signaling\n",
    "included_cell_types = [\n",
    "    # myeloid\n",
    "    \"Macrophage\",\n",
    "    \"Dendritic cell\",\n",
    "\n",
    "    # adaptive immune\n",
    "    \"B cell\",\n",
    "    \"Plasma cell\",\n",
    "    \"CD4-positive, alpha-beta T cell\",\n",
    "    \"CD8-positive, alpha-beta cytotoxic T cell\",\n",
    "    \"mature NK T cell\",\n",
    "\n",
    "    # stromal / vascular\n",
    "    \"fibroblast\",\n",
    "    \"endothelial cell\",\n",
    "    \"smooth muscle cell\"\n",
    "]\n",
    "\n",
    "# create filter_pass flag for Geneformer tokenization\n",
    "adata.obs[\"filter_pass\"] = [\n",
    "    1 if ct in included_cell_types else 0\n",
    "    for ct in adata.obs[\"cell_type_union\"]\n",
    "]\n",
    "\n",
    "# sanity check\n",
    "adata.obs[\"filter_pass\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26c35bedac4e6990",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:42.787404Z",
     "start_time": "2026-01-17T21:36:42.697176Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "os.makedirs(\"/content/input-data/\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f7fee5edb273d81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:43.594587Z",
     "start_time": "2026-01-17T21:36:42.826006Z"
    }
   },
   "outputs": [],
   "source": [
    "# Save prepared AnnData object for Geneformer tokenization\n",
    "adata.write_h5ad(\"/content/input-data/cxg_sjogrens_prepared.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b7b93bdef44b9cc8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:43.611392Z",
     "start_time": "2026-01-17T21:36:43.600555Z"
    }
   },
   "outputs": [],
   "source": [
    "custom_attrs = {\n",
    "  \"cell_type_union\": \"cell-type\", # Here, this attribute is renamed\n",
    "  \"disease\": \"disease\",\n",
    "  \"donor_id\": \"individual\", # Patient ID, later used for data splits\n",
    "  \"development_stage\": \"age\",\n",
    "  \"sex\": \"sex\"\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99a1e7175a68a27f",
   "metadata": {},
   "source": [
    "# Download Geneformer (colab installation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93bbe231ecd3fce0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T23:36:11.743108Z",
     "start_time": "2026-01-17T23:36:11.709626Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WANDB_MODE set to: disabled\n",
      "GENEFORMER_FORCE_DEVICE set to: cpu\n",
      "PYTORCH_MPS_HIGH_WATERMARK_RATIO set to: 0.0\n",
      "Prepended local geneformer path to sys.path: /Users/aumchampaneri/PycharmProjects/Geneformer/Geneformer\n",
      "Torch CUDA available: False\n",
      "MPS available: True\n",
      "Selected device for this session: mps\n"
     ]
    }
   ],
   "source": [
    "!git lfs install\n",
    "!git clone https://huggingface.co/ctheodoris/Geneformer\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46972841",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /content/Geneformer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75da6bea",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92c4ea1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensue datasets updated in Colab environment\n",
    "!pip install gcsfs==20253.0\n",
    "!pip install datasets==3.6.0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bddd2192a0a22c32",
   "metadata": {},
   "source": [
    "# Tokenize data using Geneformer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b69a7f55f883604",
   "metadata": {},
   "source": [
    "#### The single-cell transcriptomic input data is then tokenized into rank value encodings that represent a nonparametric view of the transcriptome that prioritizes genes that uniquely distinguish cell states.\n",
    "\n",
    "#### For detailed documentation of the various arguments, please see here: https://geneformer.readthedocs.io/en/latest/geneformer.tokenizer.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "da87209cc6bcb7bd",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:36:48.037778Z",
     "start_time": "2026-01-17T21:36:48.030926Z"
    }
   },
   "outputs": [],
   "source": [
    "from geneformer import TranscriptomeTokenizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1dd06886efd7ada",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:37:07.975367Z",
     "start_time": "2026-01-17T21:36:48.038459Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tokenizing /Users/aumchampaneri/PycharmProjects/Geneformer/input-data/cxg_sjogrens_prepared.h5ad\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/aumchampaneri/PycharmProjects/Geneformer/Geneformer/geneformer/tokenizer.py:544: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  for i in adata.var[\"ensembl_id_collapsed\"][coding_miRNA_loc]\n",
      "/Users/aumchampaneri/PycharmProjects/Geneformer/Geneformer/geneformer/tokenizer.py:547: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  coding_miRNA_ids = adata.var[\"ensembl_id_collapsed\"][coding_miRNA_loc]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating dataset.\n"
     ]
    }
   ],
   "source": [
    "tk = TranscriptomeTokenizer(\n",
    "    custom_attr_name_dict=custom_attrs,\n",
    "    chunk_size=512, # adjust based on available memory\n",
    "    nproc=4, # adjust based on available CPU cores\n",
    "    model_version=\"V1\"\n",
    " )\n",
    "tk.tokenize_data(\n",
    "    data_directory=\"/content/input-data/\",\n",
    "    output_directory=\"/content/processed-data/\",\n",
    "    output_prefix=\"cm_tokenized\",\n",
    "    file_format=\"h5ad\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed33b5edaf3c426b",
   "metadata": {},
   "source": [
    "# Fine-tuning: single-task"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da3d13772dbe3154",
   "metadata": {},
   "source": [
    "#### The model embedding space can be fine-tuned towards a downstream biological question, such as distinguishing healthy from disease, to enable in silico perturbation experiments predicting which perturbations would induce desired cell state shifts within the embedding space.\n",
    "\n",
    "#### For detailed documentation of the various fine-tuning strategies and related arguments, please see here: https://geneformer.readthedocs.io/en/latest/geneformer.classifier.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d35e936bfba04d96",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:37:07.989153Z",
     "start_time": "2026-01-17T21:37:07.982685Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "os.makedirs(\"/content/finetuned_models\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73812734cd7a1ee7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:37:08.021964Z",
     "start_time": "2026-01-17T21:37:07.989674Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "260117163707\n"
     ]
    }
   ],
   "source": [
    "# Generate a unique date-based identifier for each experiment\n",
    "import datetime\n",
    "\n",
    "current_date = datetime.datetime.now()\n",
    "datestamp = f\"{str(current_date.year)[-2:]}{current_date.month:02d}{current_date.day:02d}{current_date.hour:02d}{current_date.minute:02d}{current_date.second:02d}\"\n",
    "datestamp_min = f\"{str(current_date.year)[-2:]}{current_date.month:02d}{current_date.day:02d}\"\n",
    "\n",
    "output_prefix = \"cm_dz_classifier\"\n",
    "output_dir = f\"/content/finetuned_models/{datestamp}\"\n",
    "print(datestamp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4f209f2c903b505",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:37:08.029941Z",
     "start_time": "2026-01-17T21:37:08.022809Z"
    }
   },
   "outputs": [],
   "source": [
    "os.makedirs(output_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bb1a1d6e058b628",
   "metadata": {},
   "source": [
    "#### The classifier is initialized here to distinguish cell states annotated within the \"disease\" attribute. Data is filtered to only perform the fine-tuning with cells specified in the \"cell_type\" attribute.\n",
    "\n",
    "#### Please note that, as usual with deep learning models, we **highly** recommend tuning learning hyperparameters for all fine-tuning applications as this can significantly improve model performance. Example below uses previously optimized hyperparameters, but one can optimize hyperparameters with the argument n_hyperopt_trials=n in cc.validate() where n>0 and represents the number of trials for hyperparameter optimization. Importantly, the hyperparameters provided here do not represent uniformly applicable or recommended hyperparameters. They are used only as an example for quickly executing the example notebook here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "95836dd4e0acb724",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:37:08.045658Z",
     "start_time": "2026-01-17T21:37:08.030518Z"
    }
   },
   "outputs": [],
   "source": [
    "from geneformer import Classifier\n",
    "\n",
    "training_args = {\n",
    "    \"num_train_epochs\": 0.9,\n",
    "    \"learning_rate\": 0.000804,\n",
    "    \"lr_scheduler_type\": \"polynomial\",\n",
    "    \"warmup_steps\": 1812,\n",
    "    \"weight_decay\":0.258828,\n",
    "    \"per_device_train_batch_size\": 12,\n",
    "    \"seed\": 73,\n",
    "}\n",
    "\n",
    "cc = Classifier(classifier=\"cell\",\n",
    "                cell_state_dict={\"state_key\": \"disease\", \"states\": \"all\"},\n",
    "                filter_data={\"cell-type\": [\n",
    "                    \"Macrophage\",\n",
    "                    \"Dendritic cell\",\n",
    "                    \"B cell\",\n",
    "                    \"Plasma cell\",\n",
    "                    \"CD4-positive, alpha-beta T cell\",\n",
    "                    \"CD8-positive, alpha-beta cytotoxic T cell\",\n",
    "                    \"mature NK T cell\",\n",
    "                    \"fibroblast\",\n",
    "                    \"endothelial cell\",\n",
    "                    \"smooth muscle cell\"\n",
    "                ]},\n",
    "                training_args=training_args,\n",
    "                max_ncells=37000, # 36,852 cells -> Number of cells selected for training earlier in notebook - Used 2000 for debugging\n",
    "                freeze_layers=4,\n",
    "                num_crossval_splits=5,\n",
    "                forward_batch_size=32,\n",
    "                nproc=10,\n",
    "                model_version=\"V1\" # default is V2, here set to V1 model to fit into Colab 40G GPU resources\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4cc022149582b50",
   "metadata": {},
   "source": [
    "#### Split the dataset into training, validation, and test sets based on individual IDs or predefined patient groups to avoid data leakage. Balance classes across splits as possible.\n",
    "\n",
    "#### Below, predefined splits are provided, which balance potential confounding variables such as age and sex.\n",
    "\n",
    "#### (Alternatively, one can use the built-in function to split the dataset. The arguments attr_to_split and attr_to_balance are both required for this method. Set attr_to_split to the attribute on which to split data (e.g., set to “individual” to split the data by individual patients into each group). Set attr_to_balance to the list of attribute keys on which to balance data while splitting on attr_to_split (e.g., [\"age\",\"sex\"]).)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91642d4a0fe1027c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-17T21:37:08.369144Z",
     "start_time": "2026-01-17T21:37:08.047089Z"
    }
   },
   "outputs": [],
   "source": [
    "# Example: Define train/validation/test splits by donor_id\n",
    "train_ids = [\"NIDCRDC3556\",\"NIDCRDC3561\",\"NIDCRDC3617\",\"NIDCRDC3622\",\"NIDCRDC3247\",\n",
    "             \"NIDCRDC3604\",\"NIDCRDC3598\",\"NIDCRDC3632\",\"NIDCRDC3559\",\"NIDCRDC3606\",\n",
    "             \"NIDCRDC3621\",\"NIDCRDC3618\",\"NIDCRDC3630\",\"NIDCRDC3607\",\"NIDCRDC3564\",\n",
    "             \"NIDCRDC3437\"]  # ~60% of donors\n",
    "\n",
    "eval_ids = [\"NIDCRDC3560\",\"NIDCRDC3596\",\"NIDCRDC3592\"]  # ~20% of donors\n",
    "test_ids = [\"NIDCRDC3623\",\"NIDCRDC3614\",\"NIDCRDC3603\",\"NIDCRDC3633\",\"NIDCRDC3637\",\"NIDCRDC3546\"]  # ~20% of donors\n",
    "\n",
    "train_test_id_split_dict = {\n",
    "    \"attr_key\": \"individual\",\n",
    "    \"train\": train_ids + eval_ids,  # Train + eval for training\n",
    "    \"test\": test_ids\n",
    "}\n",
    "\n",
    "train_valid_id_split_dict = {\n",
    "    \"attr_key\": \"individual\",\n",
    "    \"train\": train_ids,\n",
    "    \"eval\": eval_ids\n",
    "}\n",
    "\n",
    "cc.prepare_data(input_data_file=\"/content/processed-data/cm_tokenized.dataset\",\n",
    "                output_directory=output_dir,\n",
    "                output_prefix=output_prefix,\n",
    "                split_id_dict=train_test_id_split_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c357048c32c67cbd",
   "metadata": {},
   "source": [
    "#### Perform fine-tuning with the training dataset, validating performance with held-out patients in the eval dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbcc23561e4e2e84",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2026-01-18T00:22:25.479537Z",
     "start_time": "2026-01-17T23:36:59.781263Z"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1d7f7369e83f48398c68517f92f2e683",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****** Validation split: 1/5 ******\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/aumchampaneri/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/utils/data/dataloader.py:692: UserWarning: 'pin_memory' argument is set as true but not supported on MPS now, device pinned memory won't be used.\n",
      "  warnings.warn(warn_msg)\n",
      "/Users/aumchampaneri/PycharmProjects/Geneformer/Geneformer/geneformer/collator_for_classification.py:649: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
      "  batch = {k: torch.tensor(v.clone().detach(), dtype=torch.int64) if isinstance(v, torch.Tensor) else torch.tensor(v, dtype=torch.int64) for k, v in batch.items()}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div>\n",
       "      \n",
       "      <progress value='2337' max='2337' style='width:300px; height:20px; vertical-align: middle;'></progress>\n",
       "      [2337/2337 45:19, Epoch 0/1]\n",
       "    </div>\n",
       "    <table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       " <tr style=\"text-align: left;\">\n",
       "      <th>Epoch</th>\n",
       "      <th>Training Loss</th>\n",
       "      <th>Validation Loss</th>\n",
       "      <th>Accuracy</th>\n",
       "      <th>Macro F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0.445300</td>\n",
       "      <td>0.355460</td>\n",
       "      <td>0.843309</td>\n",
       "      <td>0.812828</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "12e23cdc6feb4daaa3d82a36403db087",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/122 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "RuntimeError",
     "evalue": "Placeholder storage has not been allocated on MPS device!",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[25], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m os\u001b[38;5;241m.\u001b[39menviron[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGENEFORMER_FORCE_DEVICE\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcpu\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 3\u001b[0m all_metrics \u001b[38;5;241m=\u001b[39m \u001b[43mcc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodel_directory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mabspath\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/Users/aumchampaneri/PycharmProjects/Geneformer/Geneformer/Geneformer-V1-10M/\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;66;43;03m# set to V1 model to fit into Colab 40G GPU resources\u001b[39;49;00m\n\u001b[1;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43mprepared_input_data_file\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43moutput_dir\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43moutput_prefix\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m_labeled_train.dataset\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m    \u001b[49m\u001b[43mid_class_dict_file\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43moutput_dir\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43moutput_prefix\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m_id_class_dict.pkl\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m          \u001b[49m\u001b[43moutput_directory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mabspath\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutput_dir\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m            \u001b[49m\u001b[43moutput_prefix\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_prefix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m            \u001b[49m\u001b[43mattr_to_split\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mindividual\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m    \u001b[49m\u001b[43mattr_to_balance\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdisease\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mage\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msex\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m    \u001b[49m\u001b[43mn_hyperopt_trials\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m \t\u001b[49m\u001b[38;5;66;43;03m# Number of trials to run for hyperparameter optimization. Set it to 0 for direct training without hyperparameter optimization.\u001b[39;49;00m\n\u001b[1;32m     12\u001b[0m \u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/PycharmProjects/Geneformer/Geneformer/geneformer/classifier.py:839\u001b[0m, in \u001b[0;36mClassifier.validate\u001b[0;34m(self, model_directory, prepared_input_data_file, id_class_dict_file, output_directory, output_prefix, split_id_dict, attr_to_split, attr_to_balance, gene_balance, max_trials, pval_threshold, save_eval_output, predict_eval, predict_trainer, n_hyperopt_trials, save_gene_split_datasets, debug_gene_split_datasets)\u001b[0m\n\u001b[1;32m    836\u001b[0m         iteration_num \u001b[38;5;241m=\u001b[39m iteration_num \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    837\u001b[0m         \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[0;32m--> 839\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mevaluate_model\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    840\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtrainer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    841\u001b[0m \u001b[43m    \u001b[49m\u001b[43mnum_classes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    842\u001b[0m \u001b[43m    \u001b[49m\u001b[43mid_class_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    843\u001b[0m \u001b[43m    \u001b[49m\u001b[43meval_data\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    844\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpredict_eval\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    845\u001b[0m \u001b[43m    \u001b[49m\u001b[43mksplit_output_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    846\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_prefix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    847\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    848\u001b[0m results \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m [result]\n\u001b[1;32m    849\u001b[0m all_conf_mat \u001b[38;5;241m=\u001b[39m all_conf_mat \u001b[38;5;241m+\u001b[39m result[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mconf_mat\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[0;32m~/PycharmProjects/Geneformer/Geneformer/geneformer/classifier.py:1356\u001b[0m, in \u001b[0;36mClassifier.evaluate_model\u001b[0;34m(self, model, num_classes, id_class_dict, eval_data, predict, output_directory, output_prefix, predict_metadata)\u001b[0m\n\u001b[1;32m   1353\u001b[0m \u001b[38;5;66;03m##### Evaluate the model #####\u001b[39;00m\n\u001b[1;32m   1354\u001b[0m labels \u001b[38;5;241m=\u001b[39m id_class_dict\u001b[38;5;241m.\u001b[39mkeys()\n\u001b[0;32m-> 1356\u001b[0m y_pred, y_true, logits_list, predict_metadata_all \u001b[38;5;241m=\u001b[39m \u001b[43meu\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclassifier_predict\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1357\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclassifier\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43meval_data\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mforward_batch_size\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgene_token_dict\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpredict_metadata\u001b[49m\n\u001b[1;32m   1358\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1360\u001b[0m conf_mat, macro_f1, acc, roc_metrics \u001b[38;5;241m=\u001b[39m eu\u001b[38;5;241m.\u001b[39mget_metrics(\n\u001b[1;32m   1361\u001b[0m     y_pred, y_true, logits_list, num_classes, labels\n\u001b[1;32m   1362\u001b[0m )\n\u001b[1;32m   1363\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m predict \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n",
      "File \u001b[0;32m~/PycharmProjects/Geneformer/Geneformer/geneformer/evaluation_utils.py:130\u001b[0m, in \u001b[0;36mclassifier_predict\u001b[0;34m(model, classifier_type, evalset, forward_batch_size, gene_token_dict, predict_metadata)\u001b[0m\n\u001b[1;32m    128\u001b[0m device \u001b[38;5;241m=\u001b[39m get_device()\n\u001b[1;32m    129\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m torch\u001b[38;5;241m.\u001b[39mno_grad():\n\u001b[0;32m--> 130\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    131\u001b[0m \u001b[43m        \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minput_data_batch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    132\u001b[0m \u001b[43m        \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattn_msk_batch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    133\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlabels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlabel_batch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    134\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    135\u001b[0m     predict_logits \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m [torch\u001b[38;5;241m.\u001b[39msqueeze(outputs\u001b[38;5;241m.\u001b[39mlogits\u001b[38;5;241m.\u001b[39mto(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcpu\u001b[39m\u001b[38;5;124m\"\u001b[39m))]\n\u001b[1;32m    136\u001b[0m     predict_labels \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m [torch\u001b[38;5;241m.\u001b[39msqueeze(label_batch\u001b[38;5;241m.\u001b[39mto(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcpu\u001b[39m\u001b[38;5;124m\"\u001b[39m))]\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/module.py:1775\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1773\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1774\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1775\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/module.py:1786\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1781\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1782\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1783\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1784\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1785\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1786\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1788\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1789\u001b[0m called_always_called_hooks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m()\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/transformers/models/bert/modeling_bert.py:1482\u001b[0m, in \u001b[0;36mBertForSequenceClassification.forward\u001b[0;34m(self, input_ids, attention_mask, token_type_ids, position_ids, head_mask, inputs_embeds, labels, output_attentions, output_hidden_states, return_dict)\u001b[0m\n\u001b[1;32m   1474\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1475\u001b[0m \u001b[38;5;124;03mlabels (`torch.LongTensor` of shape `(batch_size,)`, *optional*):\u001b[39;00m\n\u001b[1;32m   1476\u001b[0m \u001b[38;5;124;03m    Labels for computing the sequence classification/regression loss. Indices should be in `[0, ...,\u001b[39;00m\n\u001b[1;32m   1477\u001b[0m \u001b[38;5;124;03m    config.num_labels - 1]`. If `config.num_labels == 1` a regression loss is computed (Mean-Square loss), If\u001b[39;00m\n\u001b[1;32m   1478\u001b[0m \u001b[38;5;124;03m    `config.num_labels > 1` a classification loss is computed (Cross-Entropy).\u001b[39;00m\n\u001b[1;32m   1479\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1480\u001b[0m return_dict \u001b[38;5;241m=\u001b[39m return_dict \u001b[38;5;28;01mif\u001b[39;00m return_dict \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconfig\u001b[38;5;241m.\u001b[39muse_return_dict\n\u001b[0;32m-> 1482\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbert\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1483\u001b[0m \u001b[43m    \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1484\u001b[0m \u001b[43m    \u001b[49m\u001b[43mattention_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mattention_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1485\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtoken_type_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtoken_type_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1486\u001b[0m \u001b[43m    \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1487\u001b[0m \u001b[43m    \u001b[49m\u001b[43mhead_mask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhead_mask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1488\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs_embeds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs_embeds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1489\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_attentions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_attentions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1490\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutput_hidden_states\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moutput_hidden_states\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1491\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_dict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1492\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1494\u001b[0m pooled_output \u001b[38;5;241m=\u001b[39m outputs[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   1496\u001b[0m pooled_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdropout(pooled_output)\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/module.py:1775\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1773\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1774\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1775\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/module.py:1786\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1781\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1782\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1783\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1784\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1785\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1786\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1788\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1789\u001b[0m called_always_called_hooks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m()\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/transformers/models/bert/modeling_bert.py:936\u001b[0m, in \u001b[0;36mBertModel.forward\u001b[0;34m(self, input_ids, attention_mask, token_type_ids, position_ids, head_mask, inputs_embeds, encoder_hidden_states, encoder_attention_mask, past_key_values, use_cache, output_attentions, output_hidden_states, return_dict, cache_position)\u001b[0m\n\u001b[1;32m    933\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    934\u001b[0m         token_type_ids \u001b[38;5;241m=\u001b[39m torch\u001b[38;5;241m.\u001b[39mzeros(input_shape, dtype\u001b[38;5;241m=\u001b[39mtorch\u001b[38;5;241m.\u001b[39mlong, device\u001b[38;5;241m=\u001b[39mdevice)\n\u001b[0;32m--> 936\u001b[0m embedding_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membeddings\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    937\u001b[0m \u001b[43m    \u001b[49m\u001b[43minput_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minput_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    938\u001b[0m \u001b[43m    \u001b[49m\u001b[43mposition_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mposition_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    939\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtoken_type_ids\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtoken_type_ids\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    940\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs_embeds\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minputs_embeds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    941\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpast_key_values_length\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpast_key_values_length\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    942\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    944\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m attention_mask \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    945\u001b[0m     attention_mask \u001b[38;5;241m=\u001b[39m torch\u001b[38;5;241m.\u001b[39mones((batch_size, seq_length \u001b[38;5;241m+\u001b[39m past_key_values_length), device\u001b[38;5;241m=\u001b[39mdevice)\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/module.py:1775\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1773\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1774\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1775\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/module.py:1786\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1781\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1782\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1783\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1784\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1785\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1786\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1788\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1789\u001b[0m called_always_called_hooks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m()\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/transformers/models/bert/modeling_bert.py:179\u001b[0m, in \u001b[0;36mBertEmbeddings.forward\u001b[0;34m(self, input_ids, token_type_ids, position_ids, inputs_embeds, past_key_values_length)\u001b[0m\n\u001b[1;32m    176\u001b[0m         token_type_ids \u001b[38;5;241m=\u001b[39m torch\u001b[38;5;241m.\u001b[39mzeros(input_shape, dtype\u001b[38;5;241m=\u001b[39mtorch\u001b[38;5;241m.\u001b[39mlong, device\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mposition_ids\u001b[38;5;241m.\u001b[39mdevice)\n\u001b[1;32m    178\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m inputs_embeds \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 179\u001b[0m     inputs_embeds \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mword_embeddings\u001b[49m\u001b[43m(\u001b[49m\u001b[43minput_ids\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    180\u001b[0m token_type_embeddings \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtoken_type_embeddings(token_type_ids)\n\u001b[1;32m    182\u001b[0m embeddings \u001b[38;5;241m=\u001b[39m inputs_embeds \u001b[38;5;241m+\u001b[39m token_type_embeddings\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/module.py:1775\u001b[0m, in \u001b[0;36mModule._wrapped_call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1773\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_compiled_call_impl(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[1;32m   1774\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1775\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_impl\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/module.py:1786\u001b[0m, in \u001b[0;36mModule._call_impl\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1781\u001b[0m \u001b[38;5;66;03m# If we don't have any hooks, we want to skip the rest of the logic in\u001b[39;00m\n\u001b[1;32m   1782\u001b[0m \u001b[38;5;66;03m# this function, and just call forward.\u001b[39;00m\n\u001b[1;32m   1783\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_forward_pre_hooks\n\u001b[1;32m   1784\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_backward_pre_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_backward_hooks\n\u001b[1;32m   1785\u001b[0m         \u001b[38;5;129;01mor\u001b[39;00m _global_forward_hooks \u001b[38;5;129;01mor\u001b[39;00m _global_forward_pre_hooks):\n\u001b[0;32m-> 1786\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mforward_call\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1788\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1789\u001b[0m called_always_called_hooks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m()\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/modules/sparse.py:192\u001b[0m, in \u001b[0;36mEmbedding.forward\u001b[0;34m(self, input)\u001b[0m\n\u001b[1;32m    191\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mforward\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;28minput\u001b[39m: Tensor) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Tensor:\n\u001b[0;32m--> 192\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mF\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membedding\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    193\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    194\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mweight\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    195\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpadding_idx\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    196\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_norm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    197\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnorm_type\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    198\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mscale_grad_by_freq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    199\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msparse\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    200\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.pyenv/versions/bioenv/lib/python3.10/site-packages/torch/nn/functional.py:2542\u001b[0m, in \u001b[0;36membedding\u001b[0;34m(input, weight, padding_idx, max_norm, norm_type, scale_grad_by_freq, sparse)\u001b[0m\n\u001b[1;32m   2536\u001b[0m     \u001b[38;5;66;03m# Note [embedding_renorm set_grad_enabled]\u001b[39;00m\n\u001b[1;32m   2537\u001b[0m     \u001b[38;5;66;03m# XXX: equivalent to\u001b[39;00m\n\u001b[1;32m   2538\u001b[0m     \u001b[38;5;66;03m# with torch.no_grad():\u001b[39;00m\n\u001b[1;32m   2539\u001b[0m     \u001b[38;5;66;03m#   torch.embedding_renorm_\u001b[39;00m\n\u001b[1;32m   2540\u001b[0m     \u001b[38;5;66;03m# remove once script supports set_grad_enabled\u001b[39;00m\n\u001b[1;32m   2541\u001b[0m     _no_grad_embedding_renorm_(weight, \u001b[38;5;28minput\u001b[39m, max_norm, norm_type)\n\u001b[0;32m-> 2542\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membedding\u001b[49m\u001b[43m(\u001b[49m\u001b[43mweight\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpadding_idx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscale_grad_by_freq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msparse\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Placeholder storage has not been allocated on MPS device!"
     ]
    }
   ],
   "source": [
    "all_metrics = cc.validate(\n",
    "    model_directory=os.path.abspath(\"/content/Geneformer/Geneformer/Geneformer-V1-10M/\"), # set to V1 model to fit into Colab 40G GPU resources\n",
    "    prepared_input_data_file=f\"{output_dir}/{output_prefix}_labeled_train.dataset\",\n",
    "    id_class_dict_file=f\"{output_dir}/{output_prefix}_id_class_dict.pkl\",\n",
    "          output_directory=os.path.abspath(output_dir),\n",
    "            output_prefix=output_prefix,\n",
    "            attr_to_split=\"individual\",\n",
    "    attr_to_balance=[\"disease\",\"age\",\"sex\"],\n",
    "    n_hyperopt_trials=0 \t# Number of trials to run for hyperparameter optimization. Set it to 0 for direct training without hyperparameter optimization.\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e95ea9f78e7829fa",
   "metadata": {},
   "source": [
    "#### Evaluate the final model on held-out patients in the test dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d4f99324796c9b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_metrics_test = cc.evaluate_saved_model(\n",
    "model_directory=f\"{output_dir}/{datestamp_min}_geneformer_cellClassifier_{output_prefix}/ksplit1/\",\n",
    "\tid_class_dict_file=f\"{output_dir}/{output_prefix}_id_class_dict.pkl\",\n",
    "     \ttest_data_file=f\"{output_dir}/{output_prefix}_labeled_test.dataset\",\n",
    "     \toutput_directory=output_dir,\n",
    "      \toutput_prefix=output_prefix,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb1504a05a0194b0",
   "metadata": {},
   "source": [
    "#### Plot confusion matrix and prediction confidence for held-out patients in the test dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb40f773895f466",
   "metadata": {},
   "outputs": [],
   "source": [
    "cc.plot_conf_mat(\n",
    "    conf_mat_dict={\"Geneformer\": all_metrics[\"conf_matrix\"]},\n",
    "    output_directory=output_dir,\n",
    "    output_prefix=output_prefix,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c02c479c037ca4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cc.plot_roc(\n",
    "    roc_metric_dict={\"Geneformer\": all_metrics[\"all_roc_metrics\"]},\n",
    "    model_style_dict={\"Geneformer\": {\"color\": \"red\", \"linestyle\": \"-\"}},\n",
    "    title=\"Dosage-sensitive vs -insensitive factors\",\n",
    "    output_directory=output_dir,\n",
    "    output_prefix=output_prefix,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77be7ebc73f40553",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74baefc2b738a362",
   "metadata": {},
   "source": [
    "# Embedding extraction and visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "236a3b46a5710543",
   "metadata": {},
   "source": [
    "#### Extract embeddings to visualize the fine-tuned embedding space.\n",
    "\n",
    "#### When extracting embeddings from the pretrained model prior to task-specific fine-tuning, setting emb_layer=-1 will extract embeddings from the second-to-last layer,  capturing more generalizable features than the last layer which is more closely linked to the pretraining objective of masked learning. However, because the fine-tuning learning objective is designed to directly reflect the biological question at hand, for the fine-tuned model, we recommend setting emb_layer=0 to extract embeddings from the final layer, which directly reflects features optimized for classification.\n",
    "\n",
    "#### For detailed documentation of the various arguments for embedding extraction, please see here: https://geneformer.readthedocs.io/en/latest/geneformer.emb_extractor.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d751f5ab0f5a3f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir /content/embs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cfd5ffb097edb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from geneformer import EmbExtractor\n",
    "\n",
    "embex = EmbExtractor(\n",
    "    model_type=\"CellClassifier\", # set to GeneClassifier or Pretrained for those model types\n",
    "    num_classes=2, # number of classes of fine-tuned model\n",
    "    filter_data={\"cell-type\":[\n",
    "        \"Macrophage\",\n",
    "        \"Dendritic cell\",\n",
    "        \"B cell\",\n",
    "        \"Plasma cell\",\n",
    "        \"CD4-positive, alpha-beta T cell\",\n",
    "        \"CD8-positive, alpha-beta cytotoxic T cell\",\n",
    "        \"mature NK T cell\",\n",
    "        \"fibroblast\",\n",
    "        \"endothelial cell\",\n",
    "        \"smooth muscle cell\"\n",
    "    ]}, # optionally can extract embeddings from a subset of cells based on the chosen attributes\n",
    "    max_ncells=37000,\n",
    "    emb_layer=0, # extracts embeddings from last layer\n",
    "    emb_label=[\"disease\", \"individual\"],\n",
    "    labels_to_plot=[\"disease\", \"individual\"],\n",
    "    forward_batch_size=128,\n",
    "    nproc=8,\n",
    "    model_version=\"V1\" # default is V2, here set to V1 model to fit into Colab 40G GPU resources\n",
    ")\n",
    "\n",
    "# Normalize and check the local model path we'll pass to the extractor\n",
    "model_path = os.path.abspath(f\"{output_dir}/{datestamp_min}_geneformer_cellClassifier_{output_prefix}/ksplit1\").rstrip('/')\n",
    "print('Will use model_path:', model_path)\n",
    "\n",
    "# Ensure cfgp is defined for downstream diagnostics even if model path missing\n",
    "cfgp = os.path.join(model_path, 'config.json')\n",
    "\n",
    "# Helpful diagnostics before calling extraction\n",
    "if os.path.exists(model_path):\n",
    "    print('Model dir exists; listing first 50 entries:')\n",
    "    try:\n",
    "        print(os.listdir(model_path)[:50])\n",
    "    except Exception as e:\n",
    "        print('Could not list MODEL_DIR contents:', e)\n",
    "    # check for key model files\n",
    "    for key in [\"config.json\", \"pytorch_model.bin\", \"model.safetensors\"]:\n",
    "        print(key, 'exists:', os.path.exists(os.path.join(model_path, key)))\n",
    "    # if config.json exists, show a short preview to help troubleshooting\n",
    "    cfgp = os.path.join(model_path, 'config.json')\n",
    "    if os.path.exists(cfgp):\n",
    "        try:\n",
    "            with open(cfgp, 'r') as _f:\n",
    "                import json\n",
    "                cfg_preview = json.load(_f)\n",
    "            print('config.json preview keys:', list(cfg_preview.keys()))\n",
    "        except Exception as e:\n",
    "            print('Could not read config.json:', e)\n",
    "else:\n",
    "    print('Model path does not exist:', model_path)\n",
    "\n",
    "# Force HF/transformers to operate offline to avoid hub validation of local paths\n",
    "os.environ['HF_HUB_OFFLINE'] = '1'\n",
    "os.environ['TRANSFORMERS_OFFLINE'] = '1'\n",
    "print('Set HF_HUB_OFFLINE and TRANSFORMERS_OFFLINE to 1')\n",
    "\n",
    "# Run embedding extraction with robust error handling to capture and print diagnostics\n",
    "try:\n",
    "    embs = embex.extract_embs(\n",
    "        model_path,\n",
    "        \"/content/processed-data/cm_tokenized.dataset\",\n",
    "        \"/content/embs\",\n",
    "        \"cm_finetuned_embs\",\n",
    "    )\n",
    "    print('Embedding extraction completed; sample shape/info:')\n",
    "    try:\n",
    "        print(getattr(embs, 'shape', None))\n",
    "    except Exception:\n",
    "        pass\n",
    "except Exception as exc:\n",
    "    import traceback\n",
    "    print('Embedding extraction failed with exception:')\n",
    "    traceback.print_exc()\n",
    "    print('\\n--- Additional MODEL_DIR diagnostics follow ---')\n",
    "    try:\n",
    "        print('Model dir:', model_path)\n",
    "        print('Contents:', os.listdir(model_path))\n",
    "    except Exception as e:\n",
    "        print('Could not list model_path contents:', e)\n",
    "    # try to surface config.json contents if present\n",
    "    try:\n",
    "        if os.path.exists(cfgp):\n",
    "            with open(cfgp, 'r') as _f:\n",
    "                import json\n",
    "                print('config.json:')\n",
    "                print(json.dumps(json.load(_f), indent=2))\n",
    "    except Exception as e:\n",
    "        print('Failed to print config.json:', e)\n",
    "    raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "607fe9b34f9a1c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot umap of cell embeddings\n",
    "embex.plot_embs(embs=embs,\n",
    "                plot_style=\"umap\",\n",
    "                output_directory=\"/content/embs\",\n",
    "                output_prefix=\"emb_umap\",\n",
    "                max_ncells_to_plot=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b25550fbe9fd8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot heatmap of cell embeddings\n",
    "embex.plot_embs(embs=embs,\n",
    "                plot_style=\"heatmap\",\n",
    "                output_directory=\"/content/embs\",\n",
    "                output_prefix=\"emb_heatmap\",\n",
    "                max_ncells_to_plot=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4ed70236be5dae6",
   "metadata": {},
   "source": [
    "# In silico perturbation to induce desired cell states"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7cb53a1faad852",
   "metadata": {},
   "source": [
    "#### The fine-tuned model is leveraged for in silico perturbation analysis to predict gene perturbations that would induce desired cell state shifts within the embedding space, for example from disease to healthy.\n",
    "\n",
    "#### In silico perturbation experiments can also be performed to predict the impact of perturbations on other genes, indicating network connections.\n",
    "\n",
    "#### The in silico perturbation strategy is versatile in terms of the perturbation type (e.g., delete, overexpress), target gene(s) to perturb individually or in combination, cells to test, model to use, and perturbation goal embedding states (e.g., source state, target state, alternate state).\n",
    "\n",
    "#### For detailed documentation of the various in silico perturbation strategies and related arguments, please see here: https://geneformer.readthedocs.io/en/latest/geneformer.in_silico_perturber.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a782ad09df0dfa1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir /content/isp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a804597c1490052",
   "metadata": {},
   "source": [
    "#### Extract embedding positions of goal state for calculating shifts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4809f6936f62d865",
   "metadata": {},
   "outputs": [],
   "source": [
    "from geneformer import EmbExtractor, InSilicoPerturber\n",
    "\n",
    "# Extract embedding positions of goal state for calculating shifts.\n",
    "# Using labels that exist in this tokenized dataset (Sjogren data) instead of the cardiomyopathy example.\n",
    "cell_states_to_model = {\n",
    "    \"state_key\": \"disease\",\n",
    "    \"start_state\": \"Sjogren syndrome\",  # cells to perturb\n",
    "    \"goal_state\": \"normal\",              # desired target state\n",
    "    \"alt_states\": []                       # no alternate states for this example\n",
    "}\n",
    "\n",
    "embex = EmbExtractor(\n",
    "    model_type=\"CellClassifier\",\n",
    "    num_classes=3, # number of classes of fine-tuned model\n",
    "    # Use the same cell types we used for fine-tuning so filtering matches this dataset\n",
    "    filter_data={\"cell-type\":[\n",
    "        \"Macrophage\",\n",
    "        \"Dendritic cell\",\n",
    "        \"B cell\",\n",
    "        \"Plasma cell\",\n",
    "        \"CD4-positive, alpha-beta T cell\",\n",
    "        \"CD8-positive, alpha-beta cytotoxic T cell\",\n",
    "        \"mature NK T cell\",\n",
    "        \"fibroblast\",\n",
    "        \"endothelial cell\",\n",
    "        \"smooth muscle cell\"\n",
    "    ]},\n",
    "     max_ncells=37000,\n",
    "     emb_mode=\"cell\",  # V1 models don't have a CLS token; use cell embeddings\n",
    "     emb_layer=0, # extracts embeddings from last layer\n",
    "     emb_label=None,  # must be None for get_state_embs\n",
    "     labels_to_plot=None,\n",
    "     forward_batch_size=128,\n",
    "     nproc=8,\n",
    "     model_version=\"V1\", # default is V2\n",
    "     summary_stat=\"exact_mean\",  # required for extracting exact state embeddings\n",
    " )\n",
    "\n",
    "# ensure isp output directory exists (prevents FileNotFoundError when saving state embs)\n",
    "import os\n",
    "os.makedirs(\"/content/isp\", exist_ok=True)\n",
    "\n",
    "state_embs_dict = embex.get_state_embs(cell_states_to_model,\n",
    "                    model_directory=f\"{output_dir}/{datestamp_min}_geneformer_cellClassifier_{output_prefix}/ksplit1/\",\n",
    "                    input_data_file=\"/content/processed-data/cm_tokenized.dataset\",\n",
    "                    output_directory=\"/content/isp\",\n",
    "                    output_prefix=output_prefix\n",
    "                    )\n",
    "\n",
    "# --- ADDED: validate and convert state_embs_dict values to torch.Tensor ---\n",
    "# This ensures downstream InSilicoPerturber receives torch tensors (required for computations).\n",
    "import torch\n",
    "try:\n",
    "    import numpy as _np\n",
    "except Exception:\n",
    "    _np = None\n",
    "\n",
    "if state_embs_dict is None:\n",
    "    raise RuntimeError('state_embs_dict returned None from embex.get_state_embs; expected dict of embeddings')\n",
    "\n",
    "for key, val in list(state_embs_dict.items()):\n",
    "    if torch.is_tensor(val):\n",
    "        continue\n",
    "    # numpy array\n",
    "    if _np is not None and isinstance(val, _np.ndarray):\n",
    "        try:\n",
    "            state_embs_dict[key] = torch.tensor(val)\n",
    "            continue\n",
    "        except Exception as e:\n",
    "            raise RuntimeError(f\"Failed to convert state_embs_dict[{key}] (ndarray) to torch.Tensor: {e}\")\n",
    "    # list/tuple or other array-like\n",
    "    try:\n",
    "        state_embs_dict[key] = torch.as_tensor(val)\n",
    "    except Exception as e:\n",
    "        raise RuntimeError(f\"state_embs_dict key '{key}' could not be converted to torch.Tensor. Provide values as torch.Tensor, numpy array, list, or pandas Series. Original error: {e}\")\n",
    "\n",
    "print('state_embs_dict validated; keys and tensor shapes:')\n",
    "for k,v in state_embs_dict.items():\n",
    "    try:\n",
    "        print(k, v.shape)\n",
    "    except Exception:\n",
    "        print(k, type(v))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91c6ce8879c281ea",
   "metadata": {},
   "source": [
    "#### Perform in silico perturbation, providing the goal embedding positions in cell_states_to_model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b01f1412f130b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbation_type = \"delete\" # or \"overexpress\"\n",
    "genes_to_perturb = \"all\"\n",
    "\n",
    "isp = InSilicoPerturber(perturb_type=perturbation_type,\n",
    "                        genes_to_perturb=genes_to_perturb,\n",
    "                        model_type=\"CellClassifier\",\n",
    "                        num_classes=3,\n",
    "                        emb_mode=\"cell\",\n",
    "                        filter_data={\"cell-type\":[\n",
    "                            \"Macrophage\",\n",
    "                            \"Dendritic cell\",\n",
    "                            \"B cell\",\n",
    "                            \"Plasma cell\",\n",
    "                            \"CD4-positive, alpha-beta T cell\",\n",
    "                            \"CD8-positive, alpha-beta cytotoxic T cell\",\n",
    "                            \"mature NK T cell\",\n",
    "                            \"fibroblast\",\n",
    "                            \"endothelial cell\",\n",
    "                            \"smooth muscle cell\"\n",
    "                        ]},\n",
    "                        cell_states_to_model=cell_states_to_model,\n",
    "                        state_embs_dict=state_embs_dict,\n",
    "                        max_ncells=100,\n",
    "                        emb_layer=0,\n",
    "                        forward_batch_size=32,  # REDUCED\n",
    "                        nproc=1,                # REDUCED\n",
    "                        model_version=\"V1\"\n",
    "                        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4b1099306fd796",
   "metadata": {},
   "outputs": [],
   "source": [
    "isp.perturb_data(model_directory = f\"{output_dir}/{datestamp_min}_geneformer_cellClassifier_{output_prefix}/ksplit1/\",\n",
    "                 input_data_file=\"/content/processed-data/cm_tokenized.dataset\",\n",
    "                 output_directory=\"/content/isp\",\n",
    "                 output_prefix=output_prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bc4ccf5a2d8476f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir /content/results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e799fc4e6d02daad",
   "metadata": {},
   "source": [
    "#### Aggregate data from in silico perturbation experiments and perform statistics.\n",
    "\n",
    "#### In this case the analysis mode is \"goal_state_shift\" given the in silico perturbation was set up to predict perturbations that shift cells towards a goal cell state.\n",
    "\n",
    "#### For detailed documentation of the various in silico perturbation stat modes, related arguments, and description of output column meanings, please see here: https://geneformer.readthedocs.io/en/latest/geneformer.in_silico_perturber_stats.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "712f79d41ed55641",
   "metadata": {},
   "outputs": [],
   "source": [
    "from geneformer import InSilicoPerturberStats\n",
    "\n",
    "cell_states_to_model = {\n",
    "    \"state_key\": \"disease\",\n",
    "    \"start_state\": \"Sjogren syndrome\",  # cells to perturb\n",
    "    \"goal_state\": \"normal\",              # desired target state\n",
    "    \"alt_states\": []                       # no alternate states for this example\n",
    "}\n",
    "\n",
    "ispstats = InSilicoPerturberStats(mode=\"goal_state_shift\",\n",
    "    genes_perturbed=\"all\",\n",
    "    combos=0,\n",
    "    anchor_gene=None,\n",
    "    cell_states_to_model=cell_states_to_model,\n",
    "    model_version=\"V1\"  # default is V2, here set to V1 model to fit into Colab 40G GPU resources\n",
    ")\n",
    "\n",
    "\n",
    "# Extracts data from intermediate files and processes stats to output in final .csv\n",
    "ispstats.get_stats(\"/content/isp\",\n",
    "                   None,\n",
    "                   \"/content/results\",\n",
    "                   \"isd_cm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "534121ff18d81730",
   "metadata": {},
   "source": [
    "#### Load output csv as dataframe to review results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d73051ec8c213f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "full_results = pd.read_csv(\"/content/results/isd_cm.csv\",index_col=0)\n",
    "full_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2de20667b887f0d2",
   "metadata": {},
   "source": [
    "#### The above in silico perturbation was only performed on 100 cells for the purposes of demonstrating the code. Below, in silico perturbation results from a larger number of cells is downloaded for visualization of results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4945193a7d276c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set_theme(style='white')\n",
    "\n",
    "# Filter for genes with significant shifts in both directions\n",
    "df = full_results[(full_results['Goal_end_FDR'] < 0.05) & (full_results['Alt_end_FDR_hcm'] < 0.05)]\n",
    "\n",
    "# Further filter for genes shifting towards one state and away from the other\n",
    "df = df[((df['Shift_to_goal_end'] > 0) & (df['Shift_to_alt_end_hcm'] < 0)) |\n",
    "        ((df['Shift_to_goal_end'] < 0) & (df['Shift_to_alt_end_hcm'] > 0))]\n",
    "\n",
    "# Filter for number of cells with that gene expressed\n",
    "df = df[df[\"N_Detections\"] > 100]\n",
    "\n",
    "n_top_genes = 15\n",
    "top_genes = df.loc[df[[\"Shift_to_goal_end\", \"Shift_to_alt_end_hcm\"]].abs().sum(axis=1).sort_values(ascending=False).index][\"Gene_name\"].tolist()[:n_top_genes]\n",
    "\n",
    "# List of genes you want to annotate\n",
    "genes_to_annotate = top_genes + [\"C3\", \"CFH\", \"C3AR1\"]\n",
    "\n",
    "# Scatter plot of shift to goal state (NF) vs shift to alternate state (HCM)\n",
    "plt.figure(figsize=(6, 6))\n",
    "sns.scatterplot(data=df,\n",
    "                x='Shift_to_goal_end',\n",
    "                y='Shift_to_alt_end_hcm',\n",
    "                color='red',  # Color for significant points with desired shifts\n",
    "                size='N_Detections',  # Size based on the number of detections\n",
    "                sizes=(20, 200),  # Adjust marker sizes\n",
    "                legend='brief')\n",
    "\n",
    "# Add titles and labels\n",
    "plt.title('Candidate therapeutic targets whose in silico repression \\n in DCM cardiomyocytes shifts towards NF, away from HCM', fontsize=16)\n",
    "plt.xlabel('Shift to goal state (NF)', fontsize=14)\n",
    "plt.ylabel('Shift to alternate state (HCM)', fontsize=14)\n",
    "\n",
    "# Add a reference line where shifts to NF and HCM are equal\n",
    "plt.axline((0, 0), slope=1, color='gray', linestyle='--')\n",
    "\n",
    "# Add red dotted lines to mark zero positions on both axes\n",
    "plt.axhline(0, color='red', linestyle=':', linewidth=1.5)\n",
    "plt.axvline(0, color='red', linestyle=':', linewidth=1.5)\n",
    "\n",
    "# # Annotate selected genes\n",
    "for _, row in df.iterrows():\n",
    "    if row['Gene_name'] in genes_to_annotate:\n",
    "        plt.text(row['Shift_to_goal_end'], row['Shift_to_alt_end_hcm'], row['Gene_name'],\n",
    "                 fontsize=10, color='black', weight='bold')\n",
    "\n",
    "# Show plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437f3f2bd55761f2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
