{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "967e4383",
   "metadata": {},
   "outputs": [],
   "source": [
    "PDIR = '/Users/aumchampaneri/VSCode Projects/complement-receptor-blockade/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6a76f84",
   "metadata": {},
   "source": [
    "# Download data from cellxgene census\n",
    "\n",
    "> Only need to run this section once for initial data downloading. Skip this section on all proceeding runs due to import incompatibility errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd0b6a99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install cellxgene-census if not already installed (run in terminal or notebook)\n",
    "# !pip install -U cellxgene-census  # Uncomment if needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a93dd59",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma(census_version=\"latest\")\n",
    "census[\"census_info\"][\"summary\"].read().concat().to_pandas()\n",
    "\n",
    "# Get all dataset citations from census\n",
    "datasets = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f75296be",
   "metadata": {},
   "source": [
    "## Fetch AnnData from Census"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86ca9434",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.makedirs(f\"{PDIR}/lupus-pbmc/raw-data\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a1219e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch an AnnData object\n",
    "dataset_id = \"218acb0f-9f2f-4f76-b90b-15a4b7c7f629\" # https://datasets.cellxgene.cziscience.com/218acb0f-9f2f-4f76-b90b-15a4b7c7f629.h5ad\n",
    "cellxgene_census.download_source_h5ad(dataset_id, to_path=f\"{PDIR}/lupus-pbmc/raw-data/cxg_lupus-pbmc.h5ad\", progress_bar=True)\n",
    "census.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79507589",
   "metadata": {},
   "source": [
    "## Fetch citation for dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5864ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a citation string for the slice\n",
    "datasets[datasets[\"dataset_id\"] == dataset_id].iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "717e7461",
   "metadata": {},
   "source": [
    "# Prepare data for Geneformer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e0bb5be",
   "metadata": {},
   "source": [
    "#### The input data for Geneformer is single-cell RNA-seq (scRNAseq) data composed of a raw count matrix stored in .loom or .h5ad format. If data is in another format (e.g., .csv, .mtx), it should be converted to .loom or .h5ad using Scanpy or Anndata tools.\n",
    "\n",
    "#### No normalization, transformation, or feature selection should be performed. Removal of cells per standard quality controls (e.g., removal of empty droplets, damaged cells, doublets, etc.) is recommended."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a55c1bd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "adata = sc.read_h5ad(f\"{PDIR}/lupus-pbmc/raw-data/cxg_lupus-pbmc.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3452c807",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78064780",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bb3e26c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40c83e0b",
   "metadata": {},
   "source": [
    "#### The genes must be labeled as Ensembl IDs. If your dataset includes only gene symbols instead, you can use external tools like Ensembl Biomart or MyGeneInfo to convert them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fc00303",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Duplicate index to a new .var column -> \"ensembl_id\"\n",
    "adata.var[\"ensembl_id\"] = adata.var.index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "703faccc",
   "metadata": {},
   "source": [
    "#### The data may contain additional feature labels that one would like to retain in the tokenized dataset, for example to use as labels for embeddings or fine-tuning, or to confirm balancing of attributes across data splits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d450bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# systemic lupus erythematosus\n",
    "# normal (control)\n",
    "set(adata.obs[\"disease\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f6a096d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cell type annotations from original authors\n",
    "set(adata.obs[\"cell_type\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5690a685",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reformat developmental stage / age attribute\n",
    "set(adata.obs[\"development_stage\"])\n",
    "\n",
    "# keep only first two characters (e.g., \"Adult\" -> \"Ad\")\n",
    "adata.obs[\"development_stage\"] = [\n",
    "    ds[:2] for ds in adata.obs[\"development_stage\"]\n",
    "]\n",
    "\n",
    "# sanity check\n",
    "set(adata.obs[\"development_stage\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e16e642",
   "metadata": {},
   "source": [
    "## Save processed AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c412ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.makedirs(f\"{PDIR}/lupus-pbmc/input-data\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43386776",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save prepared AnnData object for Geneformer tokenization\n",
    "adata.write_h5ad(f\"{PDIR}}lupus-pbmc/input-data/cxg_lupus-pbmc_prepared.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
